packages/core/src/app/client/state-manager.ts

```diff
@@ -0,0 +1,119 @@
+import BigNum = require('bn.js')
+
+import {PluginManager, PredicatePlugin, StateDB, StateManager} from "../../interfaces/client";
+import {
+  HistoryProof,
+  Range,
+  StateQuery,
+  StateQueryResult,
+  StateUpdate,
+  Transaction, VerifiedStateUpdate
+} from "../../interfaces/common/utils";
+
+/**
+ * StateManager that validates transactions and wraps and modifies StateDB as necessary.
+ *
+ * See: http://spec.plasma.group/en/latest/src/05-client-architecture/state-manager.html for more details.
+ */
+export class DefaultStateManager implements StateManager {
+
+  private stateDB: StateDB
+  private pluginManager: PluginManager
+
+  public constructor(stateDB: StateDB, pluginManager: PluginManager) {
+    this.stateDB = stateDB
+    this.pluginManager = pluginManager
+  }
+
+  public async executeTransaction(transaction: Transaction): Promise<{ stateUpdate: StateUpdate, validRanges: Range[] }> {
+    const result = {
+      stateUpdate: undefined,
+      validRanges: []
+    }
+
+    if ( ! this.validateTransaction(transaction)) {
+      return result
+    }
+
+    // TODO: Check that transaction.block > SyncManager.getLastSyncedBlock()
+
+    const {start, end}: Range = transaction.stateUpdate.id
+    const verifiedUpdates: VerifiedStateUpdate[] = await this.stateDB.getVerifiedStateUpdates(start, end)
+
+    verifiedUpdates.forEach((verifiedUpdate: VerifiedStateUpdate) => {
+      if ( ! this.validateVerifiedStateUpdate(verifiedUpdate)) {
+        // log here?
+        return
+      }
+
+      const {start: verifiedStart, end: verifiedEnd}: Range = verifiedUpdate.stateUpdate.id
+      if (transaction.block !== verifiedUpdate.verifiedBlockNumber + 1 || verifiedEnd.lte(start) || verifiedStart.gte(end)) {
+        // log here?
+        return
+      }
+
+      const predicatePlugin: PredicatePlugin = await this.pluginManager.getPlugin(verifiedUpdate.stateUpdate.newState.predicate)
+      if (!predicatePlugin) {
+        // log here?
+        return
+      }
+
+      const computedState: StateUpdate = await predicatePlugin.executeStateTransition(verifiedUpdate.stateUpdate, transaction)
+      result.validRanges.push({
+        start: BigNum.max(start, verifiedStart),
+        end: BigNum.min(end, verifiedEnd)
+      })
+
+
+      if (result.stateUpdate === undefined) {
+        result.stateUpdate = computedState
+      } else if (result.stateUpdate !== computedState) {
+        throw new Error(`State transition resulted in two different states: ${result.stateUpdate} and 
+          ${computedState}. Latter differed from former at range ${result.validRanges.pop()}.`)
+      }
+    })
+
+    return result;
+  }
+
+  public ingestHistoryProof(historyProof: HistoryProof): Promise<void> {
```

Comment:
So generally (... and @karlfloersch can comment here too since we've discussed before ...) I'm a fan of taking a snapshot of the local state when verifying history proofs instead of directly applying the proof elements to the database.

My reasoning:
1. It's much faster - you're not making repeated reads/writes.
2. It's easier to revert - you don't have to go back and erase data if the proof turns out to be invalid.
3. It's possible to check multiple proofs in parallel - you only need to "lock" a specific range when you're writing to the database after checking the proof.

Let me know what you think!

Reply:
Was just going off of the [spec](http://spec.plasma.group/en/latest/src/05-client-architecture/state-manager.html#ingesthistoryproof) here. Happy to change it how you all think is best until I'm far enough along to form my own thoughts.

Reply:
I think long term the efficiency of snapshots is unquestionably a big win. However, think it's important to keep V1 as simple & bug free as possible which means avoiding complex optimizations. That way we can use outputs from the inefficient but (hopefully) bug-free version to test the outputs of our more complex & efficient implementations.